From dcaac1bbc15d1a45d7b66396b631bf894aa6dea4 Mon Sep 17 00:00:00 2001
From: Dinesh Karthikeyan <quic_dinek@quicinc.com>
Date: Mon, 20 Mar 2023 12:06:09 +0530
Subject: [PATCH] ath12k: Add full dump support for IPQ5322

Handle the full dump collection support for IPQ5322 in MLO
capable condition.

Signed-off-by: Dinesh Karthikeyan <quic_dinek@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.c | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -20,6 +20,7 @@
 #include "wow.h"
 #include "sawf.h"
 #include "ppe.h"
+#include "ahb.h"
 
 unsigned int ath12k_debug_mask;
 module_param_named(debug_mask, ath12k_debug_mask, uint, 0644);
@@ -60,6 +61,8 @@ MODULE_PARM_DESC(mgmt_rx_reorder, "Mgmt
 static DEFINE_MUTEX(ath12k_hw_lock);
 static struct list_head ath12k_hw_groups = LIST_HEAD_INIT(ath12k_hw_groups);
 
+extern struct ath12k_coredump_info ath12k_coredump_ram_info;
+
 /* This function needs to be used only when dt has multi chip grouping information */
 static struct ath12k_hw_group *ath12k_core_hw_group_find_by_id(u8 group_id)
 {
@@ -1629,11 +1632,14 @@ static int ath12k_core_atomic_ssr_notifi
 {
 	struct ath12k_base *ab = container_of(nb, struct ath12k_base, atomic_ssr_nb);
 	struct device *dev = ab->dev;
-	bool multi_pd_arch = false;
-	phandle rproc_phandle;
-	struct device_node *rproc_node = NULL;
+	phandle rrproc_phandle;
 	struct platform_device *pdev = NULL;
+	struct device_node *rrproc_node = NULL;
 	struct platform_device *ssr_pdev = (struct platform_device *)data;
+	struct ath12k_hw_group *ag = ab->ag;
+	struct ath12k_ahb *ab_ahb = ath12k_ahb_priv(ab);
+	int ret;
+
 
 	if (event != SUBSYS_PREPARE_FOR_FATAL_SHUTDOWN)
 		return NOTIFY_DONE;
@@ -1643,33 +1649,23 @@ static int ath12k_core_atomic_ssr_notifi
 	if (!test_bit(ATH12K_FLAG_REGISTERED, &ab->dev_flags))
 		return NOTIFY_DONE;
 
-	/* Print the stats and do recover only if notification is received
-	 * for expected PD. Currently, there's only one PD for which below code
-	 * may not necessary. It's just added as a sanity check incase
-	 * core-BSP code base is modified and impacted this function.
-	 * However, below code will needed review when there will be
-	 * multi pd architecture.
-	 */
-	multi_pd_arch = of_property_read_bool(dev->of_node, "qcom,multipd_arch");
-	if (multi_pd_arch) {
-		if (of_property_read_u32(dev->of_node, "qcom,rproc",
-					 &rproc_phandle))
-			return NOTIFY_DONE;
-		rproc_node = of_find_node_by_phandle(rproc_phandle);
-		if (!rproc_node) {
-			ath12k_warn(ab, "ssr notification failed to get rproc_node\n");
-			return NOTIFY_DONE;
-		}
-		pdev = of_find_device_by_node(rproc_node);
-		if (!pdev) {
-			ath12k_warn(ab, "Failed to get pdev from device node\n");
-			return NOTIFY_DONE;
-		}
-		if (strcmp(ssr_pdev->name, pdev->name) != 0) {
-			ath12k_warn(ab, "SSR notification mismatch %s  pdev name:%s\n",
-				    (char *)data, pdev->name);
-			return NOTIFY_DONE;
-		}
+	if (of_property_read_u32(dev->of_node, "qcom,rproc_rpd",
+				 &rrproc_phandle))
+		return NOTIFY_DONE;
+	rrproc_node = of_find_node_by_phandle(rrproc_phandle);
+	if (!rrproc_node) {
+		ath12k_warn(ab, "ssr notification failed to get rproc_node\n");
+		return NOTIFY_DONE;
+	}
+	pdev = of_find_device_by_node(rrproc_node);
+	if (!pdev) {
+		ath12k_warn(ab, "Failed to get pdev from device node\n");
+		return NOTIFY_DONE;
+	}
+	if (strcmp(ssr_pdev->name, pdev->name) != 0) {
+		ath12k_warn(ab, "SSR notification mismatch %s  pdev name:%s\n",
+			    (char *)data, pdev->name);
+		return NOTIFY_DONE;
 	}
 
 	if (!(test_bit(ATH12K_FLAG_UNREGISTERING, &ab->ag->dev_flags))) {
@@ -1684,8 +1680,36 @@ static int ath12k_core_atomic_ssr_notifi
 	ath12k_qmi_free_target_mem_chunk(ab);
 
 	/* Crash the system once all the stats are dumped */
-	if (!ab->fw_recovery_support)
-		BUG_ON(1);
+	if (!ab->fw_recovery_support) {
+		if (ag->mlo_capable ) {
+			ab_ahb->tgt_rrproc->state = RPROC_CRASHED;
+			atomic_inc(&ath12k_coredump_ram_info.num_chip);
+			ath12k_info(ab,
+				    "%d chip dump collected and waiting for partner chips from ahb\n",
+				    atomic_read(&ath12k_coredump_ram_info.num_chip));
+			if (atomic_read(&ath12k_coredump_ram_info.num_chip) >= ab->ag->num_started)
+				BUG_ON(1);
+		}
+	} else {
+		if (ab_ahb->tgt_rproc) {
+			ab_ahb->tgt_rproc->state = RPROC_CRASHED;
+			ret = rproc_stop(ab_ahb->tgt_rproc, true);
+			if (ret < 0) {
+				ath12k_err(ab, "userpd rproc stop failed\n");
+				return ret;
+			}
+			ab_ahb->tgt_rproc->state = RPROC_SUSPENDED;
+		}
+		if (ab_ahb->tgt_rrproc) {
+			ab_ahb->tgt_rrproc->state = RPROC_RUNNING;
+			ret = rproc_stop(ab_ahb->tgt_rrproc, true);
+			if (ret < 0) {
+				ath12k_err(ab, "rootpd rproc stop failed\n");
+				return ret;
+			}
+			ab_ahb->tgt_rrproc->ops->coredump(ab_ahb->tgt_rrproc);
+		}
+	}
 
 	ath12k_dbg(ab, ATH12K_DBG_BOOT, "reset started\n");
 
--- a/drivers/net/wireless/ath/ath12k/ahb.c
+++ b/drivers/net/wireless/ath/ath12k/ahb.c
@@ -629,7 +629,7 @@ static int ath12k_ahb_ssr_notifier_reg(s
 {
 	struct ath12k_ahb *ab_ahb = ath12k_ahb_priv(ab);
 
-	return ath12k_rproc_register_subsys_notifier(ab_ahb->tgt_rproc->name,
+	return ath12k_rproc_register_subsys_notifier(ab_ahb->tgt_rrproc->name,
 						     &ab->ssr_nb,
 						     &ab->atomic_ssr_nb);
 }
@@ -638,7 +638,7 @@ static int ath12k_ahb_ssr_notifier_unreg
 {
 	struct ath12k_ahb *ab_ahb = ath12k_ahb_priv(ab);
 
-	return ath12k_rproc_unregister_subsys_notifier(ab_ahb->tgt_rproc->name,
+	return ath12k_rproc_unregister_subsys_notifier(ab_ahb->tgt_rrproc->name,
 						       &ab->ssr_nb,
 						       &ab->atomic_ssr_nb);
 }
@@ -664,8 +664,8 @@ static int ath12k_core_get_rproc(struct
 {
 	struct ath12k_ahb *ab_ahb = ath12k_ahb_priv(ab);
 	struct device *dev = ab->dev;
-	struct rproc *prproc;
-	phandle rproc_phandle;
+	struct rproc *prproc, *rrproc;
+	phandle rproc_phandle, rproc_rpd_node;
 
 	if (of_property_read_u32(dev->of_node, "qcom,rproc", &rproc_phandle)) {
 		ath12k_err(ab, "failed to get q6_rproc handle\n");
@@ -679,6 +679,18 @@ static int ath12k_core_get_rproc(struct
 	}
 	ab_ahb->tgt_rproc = prproc;
 
+	if (of_property_read_u32(dev->of_node, "qcom,rproc_rpd", &rproc_rpd_node)) {
+		ath12k_err(ab, "failed to get q6_root_pd_rproc handle\n");
+		return -ENOENT;
+	}
+
+	rrproc = rproc_get_by_phandle(rproc_rpd_node);
+	if (!rrproc) {
+		ath12k_err(ab, "failed to get root pd rproc\n");
+		return -EINVAL;
+	}
+	ab_ahb->tgt_rrproc = rrproc;
+
 	return 0;
 }
 
--- a/drivers/net/wireless/ath/ath12k/ahb.h
+++ b/drivers/net/wireless/ath/ath12k/ahb.h
@@ -26,7 +26,7 @@ enum ath12k_ahb_smp2p_msg_id {
 struct ath12k_base;
 
 struct ath12k_ahb {
-	struct rproc *tgt_rproc;
+	struct rproc *tgt_rproc, *tgt_rrproc;
 	struct {
 		struct device *dev;
 		struct iommu_domain *iommu_domain;
